### PR TITLE
Diviner flavor

### PR DESCRIPTION
**Pull request by BenWilson2**
## What changes are proposed in this pull request?

Adding support for diviner

## How is this patch tested?

unit tests

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Adding support for the Diviner grouped forecasting package.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -341,6 +341,8 @@
     ("py:class", "MlflowInferableDataset"),
     ("py:class", "scipy.sparse.csr.csr_matrix"),
     ("py:class", "scipy.sparse.csc.csc_matrix"),
+    ("py:class", "scipy.sparse._csr.csr_matrix"),
```
Comment (BenWilson2) [inline docs/source/conf.py:344]: Sphinx warnings on doc build for protected class non-visible.

File: docs/source/models.rst [hunk 2]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
     a non-pyfunc artifact. The output of the native ``ARIMA.predict()`` when returning confidence intervals is not
     a recognized signature type.
 
+Diviner (``diviner``) (Experimental)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The ``diviner`` model flavor enables logging of
+`diviner models <https://databricks-diviner.readthedocs.io/en/latest/index.html>`_ in MLflow format via the
```
Comment (BenWilson2) [inline docs/source/models.rst:873]: This link is broken until diviner's repo moves to 'public status'. Any day now...

File: examples/diviner/train.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    params.to_csv(f"{local}/params.csv", index=False, header=True)
+    metrics.to_csv(f"{local}/metrics.csv", index=False, header=True)
+    mlflow.log_artifacts(local, artifact_path="run_data")
+
+    # Saving the parameters and metrics as json without having to serialize to local
+    # NOTE: this requires casting of fields that cannot be serialized to JSON
+    # NOTE: Do not use both of these methods. These are shown as an either/or alternative based
+    # on how you would choose to consume, view, or analyze the per-group metrics and parameters.
+
+    params["t_scale"] = params["t_scale"].astype(str)  # NB: Cannot serialize a Timestamp as JSON
```
Comment (BenWilson2) [inline examples/diviner/train.py:90]: This is only a problem with Prophet, which is why I included this 'painful' example here. pmdarima doesn't have objects stored in the param extraction.

File: mlflow/diviner.py [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+            restricted_keys = {"n_periods", "horizon", "frequency", "groups"}
+
+            predict_conf = {k: v for k, v in conf.items() if k not in restricted_keys}
+            if not predict_groups:
+                prediction_df = self.diviner_model.predict(n_periods=n_periods, **predict_conf)
+            else:
+                prediction_df = self.diviner_model.predict_groups(
+                    groups=predict_groups, n_periods=n_periods, **predict_conf
+                )
+        else:
```
Comment (BenWilson2) [inline mlflow/diviner.py:446]: All bets are off on other libraries that may get wrapped in the future. I figured that version-locking support for implementations is the safer route to reduce user confusion.

**Comment by BenWilson2** [comment]
Note: Cross version test enablement held off until repo is public and we can test against the dev version by pulling the main branch. Will file follow-up PR when it's public to enable cross version tests.

**Comment by harupy** [comment]
Instead of updating the entire `mlflow/pypi_package_index.json`, can we just add `diviner`?

**Comment by BenWilson2** [comment]
@harupy good idea! Reverted the change and manually added diviner at the end of the file

File: docs/source/models.rst [hunk 5]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
...
+        period="240 hours",
+        initial="480 hours",
+        parallel="threads",
+        rolling_window=0.1,
+        monthly=False,
+    )
+    params.to_csv(f"{local}/params.csv", index=False, header=True)
+    metrics.to_csv(f"{local}/metrics.csv", index=False, header=True)
+
+    mlflow.log_artifacts(local, artifact_path="data")
```
Comment (harupy) [inline docs/source/models.rst:970]:
```suggestion
    with tempfile.TemporaryDirectory() as tmpdir:
        params = model.extract_model_params()
        metrics = model.cross_validate_and_score(
            horizon="72 hours",
            period="240 hours",
            initial="480 hours",
            parallel="threads",
            rolling_window=0.1,
            monthly=False,
        )
        params.to_csv(f"{tmpdir}/params.csv", index=False, header=True)
        metrics.to_csv(f"{tmpdir}/metrics.csv", index=False, header=True)

        mlflow.log_artifacts(local, artifact_path="data")
```

Can we use a temporary directory?

File: tests/diviner/test_diviner_model_export.py [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+DS_FORMAT = "%Y-%m-%dT%H:%M:%S"
+EXTRA_PYFUNC_SERVING_TEST_ARGS = [] if _is_available_on_pypi("diviner") else ["--no-conda"]
+
+
+@pytest.fixture
+def model_path(tmpdir):
+    return str(pathlib.Path(tmpdir).joinpath("model"))
+
+
+@pytest.fixture(scope="session")
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:46]:
```suggestion
@pytest.fixture(scope="module")
```

Can we set the scope to `module` because we don't need this fixture once this test module completes?

File: tests/diviner/test_diviner_model_export.py [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+def diviner_data():
+    return example_data_generator.generate_example_data(
+        column_count=3, series_count=4, series_size=365 * 4, start_dt="2019-01-01", days_period=1
+    )
+
+
+@pytest.fixture(scope="session")
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:53]:
```suggestion
@pytest.fixture(scope="module")
```

Same here (https://github.com/mlflow/mlflow/pull/5553/files#r838061013)

File: tests/diviner/test_diviner_model_export.py [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+        model_uri = f"runs:/{mlflow.active_run().info.run_id}/{artifact_path}"
+        mlflow.register_model.assert_called_once_with(
+            model_uri, "DivinerModel", await_registration_for=DEFAULT_AWAIT_MAX_SLEEP_SECONDS
+        )
+
+
+def test_pmdarima_log_model_no_registered_model_name(grouped_prophet):
+    artifact_path = "diviner"
+    register_model_patch = mock.patch("mlflow.register_model")
+    with mlflow.start_run(), register_model_patch, TempDir(chdr=True, remove_on_exit=True) as tmp:
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:294]: Can we use `tmp_path` fixture instead of `TempDir`?

File: tests/diviner/test_diviner_model_export.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+        match="The provided prediction configuration Pandas "
+        "DataFrame does not contain either the `n_periods` "
+        "or `horizon` columns.",
+    ):
+        bad_conf = pd.DataFrame({"bogus": "config"}, index=[0])
+        loaded_pyfunc_model.predict(bad_conf)
+
+    with pytest.raises(
+        MlflowException,
+        match="The `n_periods` column contains invalid data. " "Supplied type must be an integer.",
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:124]:
```suggestion
        match="The `n_periods` column contains invalid data. Supplied type must be an integer.",
```

Can we join strings?

File: docs/source/models.rst (hunk 5 above)
Comment (BenWilson2) [reply docs/source/models.rst:970]: absolutely fantastic idea! Changing.

File: tests/diviner/test_diviner_model_export.py (hunk 7 above)
Comment (BenWilson2) [reply tests/diviner/test_diviner_model_export.py:53]: Changed all 3 instances to 'module' scoping

File: tests/diviner/test_diviner_model_export.py [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+)
+
+
+DS_FORMAT = "%Y-%m-%dT%H:%M:%S"
+EXTRA_PYFUNC_SERVING_TEST_ARGS = [] if _is_available_on_pypi("diviner") else ["--no-conda"]
+
+
+@pytest.fixture
+def model_path(tmpdir):
+    return str(pathlib.Path(tmpdir).joinpath("model"))
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:44]:
```suggestion
def model_path(tmp_path):
    return tmp_path.joinpath("model")
```

I think this should work.

File: mlflow/diviner.py [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+
+        if not isinstance(n_periods, int):
+            raise MlflowException(
+                "The `n_periods` column contains invalid data. Supplied type must be an integer. "
+                f"Type supplied: {type(n_periods)}"
+            )
+
+        frequency = conf.get("frequency", None)
+
+        if self.diviner_instance_type == "GroupedProphet" and not frequency:
```
Comment (harupy) [inline mlflow/diviner.py:389]: Can we use `isinstance` here?

File: tests/diviner/test_diviner_model_export.py (hunk 10 above)
Comment (BenWilson2) [reply tests/diviner/test_diviner_model_export.py]: sure thing! I also added `test/` to .gitignore for anyone running tests locally in an IDE (since tmp_path fixture will create a cwd directory named "test/".

File: mlflow/diviner.py (hunk 11 above)
Comment (BenWilson2) [reply mlflow/diviner.py]: Yep! I had removed the logic that I was using with the instance name and forgot to convert this back to something more standard. Changed!

File: mlflow/diviner.py [hunk 12]
```diff
@@ -0,0 +1,452 @@
+import logging
```
Comment (dbczumar) [inline mlflow/diviner.py:1]: Can we add a brief module-level docstring that contains a link to the full docs on the MLflow website? Users can see this when they run `help(mlflow.diviner)`, and it might help them get started.

**Review by dbczumar — APPROVED** [review]
Long overdue LGTM! Thanks @BenWilson2 ! Tested this out manually with the examples from the Diviner repo with Prophet and pmdarima AutoARIMA, works like a charm! Let's ship it once small comments are addressed!

File: dev/run-python-flavor-tests.sh [hunk 13]
```diff
@@ -12,11 +12,14 @@ pytest tests/pyfunc --large
...
 pytest tests/utils/test_model_utils.py --large
 
-# TODO: Run tests for h2o, shap, prophet, and paddle in the cross-version-tests workflow
+# TODO: Run tests for h2o, shap, and paddle in the cross-version-tests workflow
 pytest tests/h2o --large
 pytest tests/shap --large
 pytest tests/paddle --large
 
+# TODO: When the Diviner repo is set to public status, move the tests to cross-version testing
+pytest tests/diviner --large
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply dev/run-python-flavor-tests.sh:21]: Added a 0.25 story to 3.70.

File: docs/source/models.rst [hunk 14]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
...
+You can also use the :py:func:`mlflow.diviner.load_model()` method to load MLflow Models with the ``diviner``
+model flavor in native diviner formats.
+
+Diviner Types
+~~~~~~~~~~~~~
+Diviner is a library that provides an orchestration framework for performing time series forecasting on groups of
+related series. Forecasting in ``diviner`` is accomplished through wrapping popular open source libraries such as
+`prophet <https://facebook.github.io/prophet/>`_ and `pmdarima <http://alkaline-ml.com/pmdarima/>`_. The ``diviner``
+library offers a simplified set of APIs to generate many time series forecasts from a single input DataFrame and a
+unified high-level API.
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply docs/source/models.rst:887]: great edit!

File: docs/source/models.rst [hunk 15]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
...
+
+    mlflow.log_dict(params.to_dict(), "params.json")
+    mlflow.log_dict(metrics.to_dict(), "metrics.json")
+
+Logging of the model artifact is shown in the ``pyfunc`` example below.
+
+Diviner pyfunc usage
+~~~~~~~~~~~~~~~~~~~~
+The ``diviner`` interface for the ``pyfunc`` type is a patched API that is defined by the configuration supplied
+in the argument to ``predict()``. This configuration is supplied as a single row ``Pandas DataFrame``.
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply docs/source/models.rst:1005]: much better

File: docs/source/models.rst [hunk 16]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
...
+
+As this configuration is dependent upon the underlying model type (i.e., the ``diviner.GroupedProphet.forecast()`` method has
+a different signature than does ``diviner.GroupedPmdarima.predict()``), resolution of provided argument names will be
+optimistically coerced into the appropriate argument keys.
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply docs/source/models.rst:1009]: nice change

File: docs/source/models.rst [hunk 17]
```diff
@@ -867,6 +867,198 @@ Index  yhat       yhat_lower yhat_upper
...
+
+.. note::
+    Diviner models support both "full group" and "partial group" forecasting. If a column named "groups" is present
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply docs/source/models.rst:1012]: good catch

File: examples/diviner/train.py (hunk 3 above)
Comment (BenWilson2) [reply examples/diviner/train.py:91]: a most excellent idea! 
converted in the simplest way possible (similar to what Pandas does with `.to_string()` and `.to_csv()`):

``` python
params = params.astype(dtype=str, errors="ignore")
```
to cast the entire df to str type and to ignore errors so that examples don't blow up.

File: mlflow/diviner.py (hunk 12 above)
Comment (BenWilson2) [reply mlflow/diviner.py:1]: added!

File: mlflow/diviner.py [hunk 18]
```diff
@@ -0,0 +1,452 @@
...
+    import diviner
+
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    path = pathlib.Path(path).absolute()
+
+    try:
+        path.mkdir(parents=True, exist_ok=False)
+    except FileExistsError as e:
+        raise MlflowException(f"Path '{e.filename}' already exists.")
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/diviner.py:122]: merged master and converted this logic to support the directory existence validation.

File: mlflow/diviner.py [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+            raise MlflowException(
+                "The provided prediction configuration contains both `n_periods` and `horizon` "
+                "with different values. Please provide only one of these integer values."
+            )
+        else:
+            if not n_periods and horizon:
+                n_periods = horizon
+
+        if not n_periods:
+            raise MlflowException(
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/diviner.py:377]: yep!

File: mlflow/diviner.py [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+                f"contain a `frequency` column. Configuration schema: {schema}."
+            )
+
+        predict_col = conf.get("predict_col", None)
+        predict_groups = conf.get("groups", None)
+
+        if predict_groups and not isinstance(predict_groups, List):
+            raise MlflowException(
+                "Specifying a group subset for prediction requires groups to be defined as a "
+                f"[List[Tuple|List[<group_keys>]]. Submitted group type: {type(predict_groups)}."
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/diviner.py:404]: good catch

File: .gitignore [hunk 21]
```diff
@@ -99,4 +100,4 @@ example/tutorial/R/*.nb.html
 travis_wait*.log
 
 # Pytorch logs
-lightning_logs
+lightning_logs
```
Comment (harupy) [inline .gitignore:103]: nit: can we revert this change?

File: .gitignore [hunk 22]
```diff
@@ -54,6 +54,7 @@ coverage.xml
 *.cover
 .hypothesis/
 .pytest_cache/
+test/
```
Comment (harupy) [inline .gitignore:57]: Curious why we need to ignore this directory.

**Review by harupy — COMMENTED** [review]
Left a few (non-blocking) comments, LGTM!

**Review by harupy — APPROVED** [review]

File: .gitignore (hunk 22 above)
Comment (BenWilson2) [reply .gitignore:57]: When switching over the test temporary directory generation to use the fixture `tmp_path`, the `cwd` that the test is executed from will create a directory called `test/` in that cwd. Any local run within an IDE of an isolated unit test will create that `test/` directory in a path that git will add. 
It's just to help prevent any stray local tracking server or artifact files from being added to a commit.

File: tests/diviner/test_diviner_model_export.py [hunk 23]
```diff
@@ -0,0 +1,479 @@
...
+
+    pd.testing.assert_frame_equal(grouped_pmdarima.predict(10), reloaded_model.predict(10))
+
+
+def test_diviner_log_model(grouped_prophet, tmpdir):
+
+    old_uri = mlflow.get_tracking_uri()
+    for should_start_run in [False, True]:
+        try:
+            mlflow.set_tracking_uri("test")
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:244]: @BenWilson2 Is this the line that creates a test directory?

https://github.com/mlflow/mlflow/pull/5553#discussion_r841800024

File: .gitignore (hunk 21 above)
Comment (BenWilson2) [reply .gitignore]: added the blank line back in :)
